module/VuFind/src/VuFind/Controller/HoldsController.php

```diff
@@ -102,4 +138,170 @@ public function listAction()
             ->collectRequestStats($view->recordList);
         return $view;
     }
+
+    /**
+     * Edit holds
+     *
+     * @return mixed
+     */
+    public function editAction()
+    {
+        // Stop now if the user does not have valid catalog credentials available:
+        if (!is_array($patron = $this->catalogLogin())) {
+            return $patron;
+        }
+
+        // Connect to the ILS:
+        $catalog = $this->getILS();
+
+        $holdConfig = $catalog->checkFunction('Holds', compact('patron'));
+        $details = $this->params()->fromPost('selectedIDS');
+        if (empty($holdConfig['updateFields']) || empty($details)) {
+            // Shouldn't be here. Redirect to holds
+            return $this->redirect()->toRoute('holds-list');
+        }
+        // If the user input contains a value not found in the session
+        // legal list, something has been tampered with -- abort the process.
+        if (array_diff($details, $this->holds()->getValidIds())) {
+            $this->flashMessenger()
+                ->addErrorMessage('error_inconsistent_parameters');
+            return $this->redirect()->toRoute('holds-list');
+        }
+
+        // Get list of pickup locations based on the first selected hold. This may
+        // not be perfect as pickup locations may differ per request, but it's the
+        // best we can do.
+        $holds = $catalog->getMyHolds($patron);
+        $firstDetails = reset($details);
+        $pickupLocations = [];
+        foreach ($holds as $hold) {
+            $currentUpdateDetails
+                = (string)$catalog->getUpdateHoldDetails($hold, $patron);
+            if ($currentUpdateDetails === $firstDetails) {
+                try {
+                    $pickupLocations = $catalog->getPickUpLocations($patron, $hold);
+                } catch (ILSException $e) {
+                    $this->flashMessenger()
+                        ->addErrorMessage('ils_connection_failed');
+                }
+                break;
+            }
+        }
+
+        $gatheredDetails = $this->params()->fromPost('gatheredDetails', []);
+        if ($this->params()->fromPost('updateHolds')) {
+            $validPickup = true;
```

Comment:
Might it be feasible to move all the logic that populates gatheredDetails to a separate method, again just to make this action easier to read? Not sure how woven together all the variables are, but it seems like it might be possible to split a big chunk of this processing off.

Reply:
Done.